### PR TITLE
Convert docker build to more efficient multi-stage build

### PR DESCRIPTION
**Pull request by barnybug**
The existing Docker build takes quite a while and produces a big image:
```
IMAGE               CREATED             CREATED BY                                      SIZE                COMMENT
a96932dc542c        3 minutes ago       /bin/sh -c #(nop)  EXPOSE 9000                  0B                  
cceefd88c782        3 minutes ago       /bin/sh -c #(nop)  ENTRYPOINT ["bin/imagin...   0B                  
6c4eee9ed704        3 minutes ago       /bin/sh -c go build -race -o bin/imaginary...   9.93MB              
6b3a6b0daf60        3 minutes ago       /bin/sh -c #(nop) COPY dir:c4ac65197153438...   22.9MB              
7b18991f7270        3 minutes ago       /bin/sh -c go get -u github.com/golang/dep...   38.8MB              
4d302f9420be        4 minutes ago       /bin/sh -c go get -u golang.org/x/net/context   15MB                
120c185ddbfc        4 minutes ago       /bin/sh -c #(nop) WORKDIR /go                   0B                  
04f48dafab3c        4 minutes ago       /bin/sh -c mkdir -p "$GOPATH/src" "$GOPATH...   0B                  
bd1dfce9fe59        4 minutes ago       /bin/sh -c #(nop)  ENV PATH=/go/bin:/usr/l...   0B                  
e346d6c45686        4 minutes ago       /bin/sh -c #(nop)  ENV GOPATH=/go               0B                  
11ad25abd268        4 minutes ago       /bin/sh -c curl -fsSL --insecure "$GOLANG_...   298MB               
adddd3a5bf70        5 minutes ago       /bin/sh -c #(nop)  ENV GOLANG_DOWNLOAD_SHA...   0B                  
fc4aaee59595        5 minutes ago       /bin/sh -c #(nop)  ENV GOLANG_DOWNLOAD_URL...   0B                  
4d531cc783f5        5 minutes ago       /bin/sh -c apt-get update && apt-get insta...   32MB                
874333b2c271        5 minutes ago       /bin/sh -c #(nop)  ENV GOLANG_VERSION=1.9.2     0B                  
990537a1cfbc        5 minutes ago       /bin/sh -c #(nop)  ENV PORT=9000                0B                  
e67b71924f30        5 minutes ago       /bin/sh -c apt-get update &&   DEBIAN_FRON...   594MB               
182ab48a7862        10 minutes ago      /bin/sh -c #(nop)  ENV LIBVIPS_VERSION=8.5.6    0B                  
6885853b8519        10 minutes ago      /bin/sh -c #(nop)  MAINTAINER tomas@aparic...   0B                  
dd6f76d9cc90        6 days ago          /bin/sh -c #(nop)  CMD ["/bin/bash"]            0B                  
<missing>           6 days ago          /bin/sh -c mkdir -p /run/systemd && echo '...   7B                  
<missing>           6 days ago          /bin/sh -c sed -i 's/^#\s*\(deb.*universe\...   2.76kB              
<missing>           6 days ago          /bin/sh -c rm -rf /var/lib/apt/lists/*          0B                  
<missing>           6 days ago          /bin/sh -c set -xe   && echo '#!/bin/sh' >...   745B                
<missing>           6 days ago          /bin/sh -c #(nop) ADD file:5b334adf9d9a225...   122MB               
```

This alpine based multi-stage Dockerfile produces this image:
```
IMAGE               CREATED             CREATED BY                                      SIZE                COMMENT
a16e5e859f98        4 minutes ago       /bin/sh -c #(nop)  EXPOSE 9000                  0B                  
e435bcf81b6a        4 minutes ago       /bin/sh -c #(nop)  ENTRYPOINT ["bin/imagin...   0B                  
2230c124ec2e        4 minutes ago       /bin/sh -c #(nop)  ENV PORT=9000                0B                  
203d2faf06a5        4 minutes ago       /bin/sh -c #(nop) COPY file:16d89547f80183...   9.55MB              
20c814649def        4 days ago          /bin/sh -c apk add --update --no-cache  --...   19.1MB              
72dfe9e82e16        6 days ago          /bin/sh -c #(nop)  CMD ["/bin/sh"]              0B                  
<missing>           6 days ago          /bin/sh -c #(nop) ADD file:682bfba4bfda1ed...   4.15MB         
```
(see https://docs.docker.com/engine/userguide/eng-image/multistage-build/ for more info about multi stage builds and their advantages)

The alpine build takes 35s on my desktop, the existing ubuntu one over 10minutes!

One caveat worth noting is because we're using alpine packages for libvips and golang this ties the build to the alpine releases (currently libvips 8.5.8, golang 1.9.2 - they're pretty good at staying up to date). Hopefully the speed of build and minimal size of resulting image outweigh this?

### CODE REVIEW

**Comment by h2non** [comment]
The problem with alpine is the lack of support with `glibc`. While apparently `libvips` can work with `musl`, not all its capabilities are expected to work in a reliable way, such as the libvips internal cache. This can easily generate stack overflow or other memory space related issues.

If any, I would support this as optional, alternative Docker image. May be we can also try with:
https://hub.docker.com/r/frolvlad/alpine-glibc/

Ideas?

**Comment by barnybug** [comment]
Thanks for the feedback - I hadn't appreciated the implications of `musl`.
I've instead redone it as a multi-stage using ubuntu, now the image becomes:
```
IMAGE               CREATED              CREATED BY                                      SIZE                COMMENT
49de7d8ad345        About a minute ago   /bin/sh -c #(nop)  EXPOSE 9000                  0B                  
98cacd2f9706        About a minute ago   /bin/sh -c #(nop)  ENTRYPOINT ["bin/imagin...   0B                  
9daa564114b2        About a minute ago   /bin/sh -c #(nop)  ENV PORT=9000                0B                  
6fa72cc73bc3        About a minute ago   /bin/sh -c #(nop) COPY file:c06b757270d99a...   9.93MB              
f0c9568560c2        About a minute ago   /bin/sh -c ldconfig                             27.8kB              
548fb05f0dd6        About a minute ago   /bin/sh -c #(nop) COPY dir:d1408600ae39db4...   41.5MB              
fb7c424e2334        About a minute ago   /bin/sh -c apt-get update &&   DEBIAN_FRON...   85.4MB              
dd6f76d9cc90        10 days ago          /bin/sh -c #(nop)  CMD ["/bin/bash"]            0B                  
<missing>           10 days ago          /bin/sh -c mkdir -p /run/systemd && echo '...   7B                  
<missing>           10 days ago          /bin/sh -c sed -i 's/^#\s*\(deb.*universe\...   2.76kB              
<missing>           10 days ago          /bin/sh -c rm -rf /var/lib/apt/lists/*          0B                  
<missing>           10 days ago          /bin/sh -c set -xe   && echo '#!/bin/sh' >...   745B                
<missing>           10 days ago          /bin/sh -c #(nop) ADD file:5b334adf9d9a225...   122MB               
```

Which is still a nice saving of ~870MB. :-)

**Comment by jbergstroem** [comment]
@barnybug there are ways of adding glibc to alpine then linking against it that aren't too hard to follow/get done. I think another approach could be just building on top of a busybox/glibc based toolchain statically and chuck the resulting binary in a container.

**Comment by tillkruss** [comment]
If `alpine-glibc` could be used that would be fantastic. As a first time user, I was surprised how long the build took and how big it was.

**Comment by barnybug** [comment]
I gave it a shot, but was getting obscure errors from golang relating to include headers - didn't have the time to venture further into why.

The approach I settled on using multistage still offers a good improvement on size.

Sure a future PR could further improve :-)

**Comment by Dynom** [comment]
We could give https://github.com/docker-slim/docker-slim a go

**Comment by Dynom** [comment]
This ticket has relevance to: https://github.com/h2non/imaginary/issues/125

**Comment by Dynom** [comment]
I don't think this PR is still relevant. We currently use a multi-stage Dockerfile, so I'll close it for now.

Thanks for your contribution however!
